1.6/Source/PawnStorages/PawnStoragesModule-VEF/CompPipedFarmNutrition.cs

```diff
@@ -1,55 +1,100 @@
 {
-    public CompResource _compResource;
-    public bool _haveCheckedForComp = false;
-    public virtual CompResource compResource
-    {
-        get
-        {
-            if (!_haveCheckedForComp && _compResource == null && parent.HasComp<CompResource>())
-            {
-                _compResource = parent.GetComp<CompResource>();
-                // make sure we only check once.
-                _haveCheckedForComp = true;
-            }
-            return _compResource;
-        }
-    }
+    private Lazy<CompResourceStorage> _resourceStorage;
 
-    public override bool IsPiped => IsAttachedToNet(out PipeNet pipeNet, out CompResource resource);
-    public override float storedNutrition => IsAttachedToNet(out PipeNet pipeNet, out CompResource resource) ? pipeNet.Stored : base.storedNutrition;
-    public override float MaxNutrition => IsAttachedToNet(out PipeNet pipeNet, out CompResource resource) ? pipeNet.AvailableCapacity : base.MaxNutrition;
+    public virtual CompResourceStorage ResourceStorage => _resourceStorage.Value;
+    public PipeNet PipeNet => ResourceStorage?.PipeNet;
+    public override bool IsPiped => true;
+    public override float StoredNutrition => PipeNet?.Stored ?? 0;
+    public override float MaxNutrition => PipeNet?.AvailableCapacity ?? 0;
 
-    public bool IsAttachedToNet(out PipeNet pipeNet, out CompResource resource)
+    public override void Initialize(CompProperties properties)
     {
-        pipeNet = null;
-        resource = compResource;
-        if (resource is not { PipeNet: { } net }) return false;
-        pipeNet = net;
-        return pipeNet.connectors.Count > 1;
+        base.Initialize(properties);
+        _resourceStorage = new Lazy<CompResourceStorage>(() => parent?.GetComp<CompResourceStorage>());
     }
 
     public override bool AbsorbToFeedIfNeeded(Need_Food foodNeeds, float desiredFeed, out float amountFed)
     {
-        if (!IsAttachedToNet(out PipeNet pipeNet, out CompResource resource)) return base.AbsorbToFeedIfNeeded(foodNeeds, desiredFeed, out amountFed);
+        /*
+         * Priority:
+         * 1) pull from internal storage
+         * 2) pull from the network
+         * 3) pull from hoppers
+         */
+        amountFed = 0;
+        if (PipeNet == null || ResourceStorage == null)
+            return false;
+
+        // Try to absorb from the network if needed
+        if (desiredFeed > ResourceStorage.AmountStored)
+        {
+            // Try to pull what we need to make up the desire
+            float toPull = desiredFeed - ResourceStorage.AmountStored;
+            PipeNet.DrawAmongStorage(toPull, PipeNet.storages.Except(ResourceStorage).ToList());
+            AddNutritionToStorage(toPull);
+        }
+
+        // Pull from storage, and update needs accordingly
+        amountFed = Mathf.Min(ResourceStorage.AmountStored, desiredFeed);
+        foodNeeds.CurLevel += amountFed;
+        ResourceStorage.DrawResource(amountFed);
```

Comment:
might be a subtle multithreading bug here. There's a minor above where you add to storage then amount fed could be less but that wouldn't matter too much. This one though might be more of an issue because if something else draws the nutrition away before you grab it this might go below 0 which we think is the trigger for the crash.

Reply:
I don't think it's the source of a crash, both `CurLevel` and `DrawResource` have guards. VEF logs an error, but just sets it to zero.

```c#
  public virtual float CurLevel
  {
    get => this.curLevelInt;
    set => this.curLevelInt = Mathf.Clamp(value, 0.0f, this.MaxLevel);
  }
```

```c#
        public void DrawResource(float amount)
        {
            amountStored -= amount;
            if (amountStored < 0f) // If we withdrawn to much, error and set stored amount to 0
            {
                Log.Error("[PipeSystem] Drawing resource we don't have from " + parent);
                amountStored = 0f;
            }
        }
```